### PR TITLE
ekf: change gyro & accel dt from float to uint64.

### PR DESCRIPTION
**Pull request by nicolaerosia**
This has the benefit of calculating the estimator timeslip correctly.

Signed-off-by: Nicolae Rosia <nicolae.rosia@gmail.com>

### CODE REVIEW

**Comment by nicolaerosia** [comment]
before: http://logs.px4.io/plot_app?log=dbf660a3-e846-4606-92aa-24ba38269eb5
after: http://logs.px4.io/plot_app?log=3605731a-7cf7-4b0f-babf-ae8f204a5dcc

File: src/modules/ekf2/ekf2_main.cpp [hunk 1]
```diff
@@ -591,14 +591,16 @@ void Ekf2::task_main()
...
 		float accel_integral[3];
-		accel_integral[0] = sensors.accelerometer_m_s2[0] * sensors.accelerometer_integral_dt;
-		accel_integral[1] = sensors.accelerometer_m_s2[1] * sensors.accelerometer_integral_dt;
-		accel_integral[2] = sensors.accelerometer_m_s2[2] * sensors.accelerometer_integral_dt;
-		_ekf.setIMUData(now, sensors.gyro_integral_dt * 1.e6f, sensors.accelerometer_integral_dt * 1.e6f,
+		float accel_dt = sensors.accelerometer_integral_dt / 1.e6f;
+		accel_integral[0] = sensors.accelerometer_m_s2[0] * accel_dt;
+		accel_integral[1] = sensors.accelerometer_m_s2[1] * accel_dt;
+		accel_integral[2] = sensors.accelerometer_m_s2[2] * accel_dt;
+		_ekf.setIMUData(now, sensors.gyro_integral_dt, sensors.accelerometer_integral_dt,
```
Comment (dagar) [inline src/modules/ekf2/ekf2_main.cpp]: @nicolaerosia this changes accel_integral sent to ecl.

**Comment by LorenzMeier** [comment]
@nicolaerosia Great work! I found a couple of locations that still might need replacement. Would you mind doing these as well?

```
Searching 2605 files for "gyro_integral_dt"

src/examples/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp:
 1395  #if 0
 1396  
 1397: 	float deltaTIntegral = _sensor_combined.gyro_integral_dt;
 1398  	float deltaTIntAcc = _sensor_combined.accelerometer_integral_dt;
 1399  

src/lib/ecl/matlab/EKF_replay/Common/convert_px4_sensor_combined_csv_data.m:
   17  n_samples = length(timestamp);
   18  imu_data.time_us = timestamp(2:n_samples) + accelerometer_timestamp_relative(2:n_samples);
   19: imu_data.gyro_dt = gyro_integral_dt(2:n_samples);
   20  imu_data.del_ang = 0.5 * ([gyro_rad0(1:n_samples-1).*imu_data.gyro_dt, ...
   21      gyro_rad1(1:n_samples-1).*imu_data.gyro_dt, ...
```

**Comment by nicolaerosia** [comment]
@LorenzMeier thanks! the example is already in this PR. The ECL changes are here: https://github.com/PX4/ecl/pull/300

File: msg/ekf2_replay.msg [hunk 2]
```diff
@@ -1,6 +1,6 @@
 
-float32 gyro_integral_dt		# gyro integration period in s
-float32 accelerometer_integral_dt 	# accelerometer integration period in s
+uint64 gyro_integral_dt		# gyro integration period in s
```
Comment (priseborough) [inline msg/ekf2_replay.msg]: Wrong units specified in comment. This happens in other places in code.

File: msg/ekf2_replay.msg [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
+uint64 accelerometer_integral_dt 	# accelerometer integration period in s
```
Comment (priseborough) [inline msg/ekf2_replay.msg]: Change units to uSec.

File: msg/sensor_combined.msg [hunk 4]
```diff
@@ -10,11 +10,11 @@ int32 RELATIVE_TIMESTAMP_INVALID = 2147483647 # (0x7fffffff) If one of the relat
 
 # gyro timstamp is equal to the timestamp of the message
 float32[3] gyro_rad			# average angular rate measured in the XYZ body frame in rad/s over the last gyro sampling period
-float32 gyro_integral_dt		# gyro measurement sampling period in s
+uint64 gyro_integral_dt		# gyro measurement sampling period in s
```
Comment (priseborough) [inline msg/sensor_combined.msg]: Change units to uSec.

File: msg/sensor_combined.msg [hunk 5]
```diff
@@ -10,11 +10,11 @@ int32 RELATIVE_TIMESTAMP_INVALID = 2147483647 # (0x7fffffff) If one of the relat
...
 
 int32 accelerometer_timestamp_relative	# timestamp + accelerometer_timestamp_relative = Accelerometer timestamp
 float32[3] accelerometer_m_s2		# average value acceleration measured in the XYZ body frame in m/s/s over the last accelerometer sampling period
-float32 accelerometer_integral_dt	# accelerometer measurement sampling period in s
+uint64 accelerometer_integral_dt	# accelerometer measurement sampling period in s
```
Comment (priseborough) [inline msg/sensor_combined.msg]: Change units to uSec.

**Comment by nicolaerosia** [comment]
@priseborough could you specify where? I think I already did that in https://github.com/PX4/Firmware/pull/7602/commits/be5842e78b92749f1a8b509a624db64509e5b4a8

**Comment by priseborough** [comment]
OK you already did. I'll look at the ecl PR now.

**Comment by priseborough** [comment]
I have just realised a significant problem with this change. Doesn't it mean that we can no longer replay logs generated using previous firmware versions?

**Comment by LorenzMeier** [comment]
Yes it does. We should consider extending the field names with _us and see if we can make the replay infrastructure auto-detect it @bkueng

**Comment by nicolaerosia** [comment]
@priseborough do you usually replay old logs on new firmware? When is it acceptable to introduce breaking changes with regards to messages?

**Comment by bkueng** [comment]
@nicolaerosia I'll add a compatibility transformation, but such cases should really be the exception.

**Comment by bkueng** [comment]
@nicolaerosia can you merge https://github.com/PX4/Firmware/tree/replay_compat into this?
It adds the compatibility. @priseborough can you validate?

**Comment by nicolaerosia** [comment]
@bkueng done

**Comment by priseborough** [comment]
@bkueng I was able to replay the following legacy log successfully.

http://review.px4.io/plot_app?log=ed369adf-4e46-486a-bc18-14a2ed188bd0

**Comment by nicolaerosia** [comment]
I had to rebase

**Pull request merged**